File: swift/Sources/NXS/NXSWriter.swift
// NXS Writer — direct-to-buffer .nxb emitter for Swift 5.9+.
//
// Mirrors the Rust NxsWriter API:
//   NXSSchema — precompile keys once; share across NXSWriter instances.
//   NXSWriter — slot-based hot path; no per-key dictionary lookups during write.
//
// Usage:
//   import NXS
//
//   let schema = NXSSchema(keys: ["id", "username", "score", "active"])
//   let w = NXSWriter(schema: schema)
//   w.beginObject()
//   w.writeI64(slot: 0, value: 42)
//   w.writeStr(slot: 1, value: "alice")
//   w.writeF64(slot: 2, value: 9.5)
//   w.writeBool(slot: 3, value: true)
//   w.endObject()
//   let bytes: [UInt8] = w.finish()

import Foundation

// ── MurmurHash3-64 ────────────────────────────────────────────────────────────

private func murmur3_64(_ data: [UInt8]) -> UInt64 {
    let C1: UInt64 = 0xFF51AFD7ED558CCD
    let C2: UInt64 = 0xC4CEB9FE1A85EC53
    var h: UInt64 = 0x93681D6255313A99
    let len = data.count
    var i = 0
    while i < len {
        var k: UInt64 = 0
        for b in 0..<8 where i + b < len {
            k |= UInt64(data[i + b]) << (b * 8)
        }
        k = k &* C1; k ^= k >> 33
        h ^= k
        h = h &* C2; h ^= h >> 33
        i += 8
    }
    h ^= UInt64(len); h ^= h >> 33
    h = h &* C1;      h ^= h >> 33
    return h
}

// ── Schema ────────────────────────────────────────────────────────────────────

public final class NXSSchema {
    public let keys: [String]
    let bitmaskBytes: Int

    public init(keys: [String]) {
        self.keys = keys
        self.bitmaskBytes = max(1, (keys.count + 6) / 7)
    }

    var count: Int { keys.count }
}

// ── Frame ─────────────────────────────────────────────────────────────────────

private struct Frame {
    let start: Int
    var bitmask: [UInt8]
    var offsetTable: [Int]      // relative offsets, write order
    var slotOffsets: [(Int, Int)] // (slot, bufPos)
    var lastSlot: Int = -1
    var needsSort: Bool = false

    init(start: Int, bitmaskBytes: Int) {
        self.start = start
        self.bitmask = [UInt8](repeating: 0, count: bitmaskBytes)
        for i in 0..<(bitmaskBytes - 1) { self.bitmask[i] = 0x80 }
        self.offsetTable = []
        self.slotOffsets = []
    }
}

// ── Writer ────────────────────────────────────────────────────────────────────

// ── Sigil constants ────────────────────────────────────────────────────────────

private let SIGIL_STR:    UInt8 = 0x22 // '"' — string / var-length
private let SIGIL_I64:    UInt8 = 0x69 // 'i'
private let SIGIL_F64:    UInt8 = 0x64 // 'd'
private let SIGIL_BOOL:   UInt8 = 0x62 // 'b'
private let SIGIL_NULL:   UInt8 = 0x6E // 'n'
private let SIGIL_BINARY: UInt8 = 0x42 // 'B'

public final class NXSWriter {
    private let schema: NXSSchema
    private var buf: [UInt8] = []
    private var frames: [Frame] = []
    private var recordOffsets: [Int] = []
    // Sigil per slot: default str/var-length; updated on each typed write
    private var slotSigils: [UInt8]

    public init(schema: NXSSchema) {
        self.schema = schema
        self.slotSigils = [UInt8](repeating: SIGIL_STR, count: schema.count)
        buf.reserveCapacity(4096)
    }

    public func beginObject() {
        if frames.isEmpty { recordOffsets.append(buf.count) }
        let start = buf.count
        let frame = Frame(start: start, bitmaskBytes: schema.bitmaskBytes)

        appendU32(0x4E59584F)   // NYXO
        appendU32(0)            // length placeholder
        buf.append(contentsOf: frame.bitmask)
        buf.append(contentsOf: [UInt8](repeating: 0, count: schema.count * 2))
        while (buf.count - start) % 8 != 0 { buf.append(0) }

        frames.append(frame)
    }

    public func endObject() {
        precondition(!frames.isEmpty, "endObject without beginObject")
        let frame = frames.removeLast()
        let totalLen = buf.count - frame.start

        // Back-patch Length
        withUInt32LE(UInt32(totalLen)) { buf[$0 + frame.start + 4] = $1 }

        // Back-patch bitmask
        let bmOff = frame.start + 8
        for (i, b) in frame.bitmask.enumerated() { buf[bmOff + i] = b }

        // Back-patch offset table
        let otStart = bmOff + schema.bitmaskBytes
        let present = frame.offsetTable.count

        if !frame.needsSort {
            for (i, rel) in frame.offsetTable.enumerated() {
                let v = UInt16(rel)
                buf[otStart + i * 2]     = UInt8(v & 0xFF)
                buf[otStart + i * 2 + 1] = UInt8(v >> 8)
            }
        } else {
            let sorted = frame.slotOffsets.sorted { $0.0 < $1.0 }
            for (i, (_, bufPos)) in sorted.enumerated() {
                let v = UInt16(bufPos - frame.start)
                buf[otStart + i * 2]     = UInt8(v & 0xFF)
                buf[otStart + i * 2 + 1] = UInt8(v >> 8)
            }
        }
        for i in (present * 2)..<(schema.count * 2) { buf[otStart + i] = 0 }
    }

    public func finish() -> [UInt8] {
        precondition(frames.isEmpty, "unclosed objects")

        let schemaBytes = buildSchemaBytes()
        let dictHash    = murmur3_64(schemaBytes)
        let dataStart   = 32 + schemaBytes.count

        let dataSector  = buf
        let tailPtr     = UInt64(dataStart + dataSector.count)
        let tail        = buildTailIndex(dataStart: dataStart, tailPtr: tailPtr)

        var out = [UInt8]()
        out.reserveCapacity(32 + schemaBytes.count + dataSector.count + tail.count)

        appendU32To(&out, 0x4E595842)         // NYXB
        appendU16To(&out, 0x0101)             // VERSION
        appendU16To(&out, 0x0002)             // FLAG_SCHEMA_EMBEDDED
        appendU64To(&out, dictHash)
        appendU64To(&out, 0)
        out.append(contentsOf: [UInt8](repeating: 0, count: 8)) // reserved

        out.append(contentsOf: schemaBytes)
        out.append(contentsOf: dataSector)
        out.append(contentsOf: tail)
        return out
    }

    // ── Typed write methods ────────────────────────────────────────────────────

    public func writeI64(slot: Int, value: Int64) {
        slotSigils[slot] = SIGIL_I64
        markSlot(slot)
        appendI64(value)
    }

    public func writeF64(slot: Int, value: Double) {
        slotSigils[slot] = SIGIL_F64
        markSlot(slot)
        appendF64(value)
    }

    public func writeBool(slot: Int, value: Bool) {
        slotSigils[slot] = SIGIL_BOOL
        markSlot(slot)
        buf.append(value ? 0x01 : 0x00)
        buf.append(contentsOf: [UInt8](repeating: 0, count: 7))
    }

    public func writeTime(slot: Int, unixNs: Int64) {
        slotSigils[slot] = SIGIL_I64
        writeI64(slot: slot, value: unixNs)
    }

    public func writeNull(slot: Int) {
        slotSigils[slot] = SIGIL_NULL
        markSlot(slot)
        buf.append(contentsOf: [UInt8](repeating: 0, count: 8))
    }

    public func writeStr(slot: Int, value: String) {
        slotSigils[slot] = SIGIL_STR
        markSlot(slot)
        let bytes = Array(value.utf8)
        appendU32(UInt32(bytes.count))
        buf.append(contentsOf: bytes)
        let used = (4 + bytes.count) % 8
        if used != 0 { buf.append(contentsOf: [UInt8](repeating: 0, count: 8 - used)) }
    }

    public func writeBytes(slot: Int, value: [UInt8]) {
        slotSigils[slot] = SIGIL_BINARY
        markSlot(slot)
        appendU32(UInt32(value.count))
        buf.append(contentsOf: value)
        let used = (4 + value.count) % 8
        if used != 0 { buf.append(contentsOf: [UInt8](repeating: 0, count: 8 - used)) }
    }

    public func writeListI64(slot: Int, values: [Int64]) {
        markSlot(slot) // list is var-length — keep SIGIL_STR default
        let total = 16 + values.count * 8
        appendU32(0x4E59584C)                  // NYXL
        appendU32(UInt32(total))
        buf.append(0x3D)                       // '=' sigil
        appendU32(UInt32(values.count))
        buf.append(contentsOf: [0, 0, 0])
        for v in values { appendI64(v) }
    }

    public func writeListF64(slot: Int, values: [Double]) {
        markSlot(slot)
        let total = 16 + values.count * 8
        appendU32(0x4E59584C)
        appendU32(UInt32(total))
        buf.append(0x7E)                       // '~' sigil
        appendU32(UInt32(values.count))
        buf.append(contentsOf: [0, 0, 0])
        for v in values { appendF64(v) }
    }

    // Convenience: write records from [[String: Any]].
    public static func fromRecords(keys: [String],
                                   records: [[String: Any]]) -> [UInt8] {
        let schema = NXSSchema(keys: keys)
        let w = NXSWriter(schema: schema)
        for rec in records {
            w.beginObject()
            for (i, key) in keys.enumerated() {
                guard let val = rec[key] else { continue }
                switch val {
                case is NSNull:        w.writeNull(slot: i)
                case let v as Bool:    w.writeBool(slot: i, value: v)
                case let v as Int:     w.writeI64(slot: i, value: Int64(v))
                case let v as Int64:   w.writeI64(slot: i, value: v)
                case let v as Double:  w.writeF64(slot: i, value: v)
                case let v as String:  w.writeStr(slot: i, value: v)
                default: break
                }
            }
            w.endObject()
        }
        return w.finish()
    }

    // ── Private helpers ────────────────────────────────────────────────────────

    private func markSlot(_ slot: Int) {
        precondition(!frames.isEmpty, "write outside beginObject/endObject")
        let byteIdx = slot / 7
        let bitIdx  = slot % 7
        frames[frames.count - 1].bitmask[byteIdx] |= UInt8(1 << bitIdx)

        let rel = buf.count - frames[frames.count - 1].start
        if slot < frames[frames.count - 1].lastSlot {
            frames[frames.count - 1].needsSort = true
        }
        frames[frames.count - 1].lastSlot = slot
        frames[frames.count - 1].offsetTable.append(rel)
        frames[frames.count - 1].slotOffsets.append((slot, buf.count))
    }

    private func appendU32(_ v: UInt32) {
        buf.append(UInt8(v & 0xFF)); buf.append(UInt8((v >> 8) & 0xFF))
        buf.append(UInt8((v >> 16) & 0xFF)); buf.append(UInt8((v >> 24) & 0xFF))
    }

    private func appendU16(_ v: UInt16) {
        buf.append(UInt8(v & 0xFF)); buf.append(UInt8((v >> 8) & 0xFF))
    }

    private func appendI64(_ v: Int64) {
        let u = UInt64(bitPattern: v)
        for i in 0..<8 { buf.append(UInt8((u >> (i * 8)) & 0xFF)) }
    }

    private func appendF64(_ v: Double) {
        let u = v.bitPattern
        for i in 0..<8 { buf.append(UInt8((u >> (i * 8)) & 0xFF)) }
    }

    private func appendU32To(_ arr: inout [UInt8], _ v: UInt32) {
        arr.append(UInt8(v & 0xFF)); arr.append(UInt8((v >> 8) & 0xFF))
        arr.append(UInt8((v >> 16) & 0xFF)); arr.append(UInt8((v >> 24) & 0xFF))
    }

    private func appendU16To(_ arr: inout [UInt8], _ v: UInt16) {
        arr.append(UInt8(v & 0xFF)); arr.append(UInt8((v >> 8) & 0xFF))
    }

    private func appendU64To(_ arr: inout [UInt8], _ v: UInt64) {
        for i in 0..<8 { arr.append(UInt8((v >> (i * 8)) & 0xFF)) }
    }

    private func withUInt32LE(_ v: UInt32, _ body: (Int, UInt8) -> Void) {
        for i in 0..<4 { body(i, UInt8((v >> (i * 8)) & 0xFF)) }
    }

    private func buildSchemaBytes() -> [UInt8] {
        let n    = schema.count
        let utf8 = schema.keys.map { Array($0.utf8) }
        var size = 2 + n + utf8.reduce(0) { $0 + $1.count + 1 }
        let pad  = (8 - size % 8) % 8
        size += pad

        var b = [UInt8](repeating: 0, count: size)
        var p = 0
        b[p] = UInt8(n & 0xFF); p += 1
        b[p] = UInt8((n >> 8) & 0xFF); p += 1
        for i in 0..<n { b[p] = slotSigils[i]; p += 1 }
        for e in utf8 {
            for byte in e { b[p] = byte; p += 1 }
            b[p] = 0; p += 1
        }
        return b
    }

    private func buildTailIndex(dataStart: Int, tailPtr: UInt64) -> [UInt8] {
        let nr = recordOffsets.count
        var b  = [UInt8]()
        b.reserveCapacity(4 + nr * 10 + 12)

        appendU32To(&b, UInt32(nr))
        for (i, rel) in recordOffsets.enumerated() {
            let lo = UInt16(i & 0xFFFF)
            b.append(UInt8(lo & 0xFF)); b.append(UInt8(lo >> 8))
            let abs = UInt64(dataStart + rel)
            appendU64To(&b, abs)
        }
        appendU64To(&b, tailPtr)
        appendU32To(&b, 0x2153584E) // NXS!
        return b
    }
}

File: kotlin/src/main/kotlin/nxs/NxsWriter.kt
package nxs

import java.io.ByteArrayOutputStream

/**
 * NXS Writer — direct-to-buffer .nxb emitter for Kotlin/JVM.
 *
 * Mirrors the Rust NxsWriter API:
 * - [NxsSchema] — precompile keys once; share across [NxsWriter] instances.
 * - [NxsWriter] — slot-based hot path; no per-key map lookups during write.
 *
 * Usage:
 * ```
 * val schema = NxsSchema(listOf("id", "username", "score", "active"))
 * val w = NxsWriter(schema)
 * w.beginObject()
 * w.writeI64(0, 42L)
 * w.writeStr(1, "alice")
 * w.writeF64(2, 9.5)
 * w.writeBool(3, true)
 * w.endObject()
 * val bytes: ByteArray = w.finish()
 * ```
 */

private fun murmurHash64(data: ByteArray): Long {
    val c1 = -0xAE502812AA7333L // 0xFF51AFD7ED558CCD as Long
    val c2 = -0x3B314601E57A13ADL // 0xC4CEB9FE1A85EC53 as Long
    var h = -0x6C97E29DAACEC567L // 0x93681D6255313A99 as Long
    val len = data.size
    var i = 0
    while (i < len) {
        var k = 0L
        for (b in 0..7) {
            if (i + b < len) k = k or ((data[i + b].toLong() and 0xFF) shl (b * 8))
        }
        k *= c1
        k = k xor (k ushr 33)
        h = h xor k
        h *= c2
        h = h xor (h ushr 33)
        i += 8
    }
    h = h xor len.toLong()
    h = h xor (h ushr 33)
    h *= c1
    h = h xor (h ushr 33)
    return h
}

class NxsSchema(val keys: List<String>) {
    val bitmaskBytes: Int = maxOf(1, (keys.size + 6) / 7)
    val count: Int get() = keys.size
}

private class Frame(val start: Int, bitmaskBytes: Int) {
    val bitmask: ByteArray =
        ByteArray(bitmaskBytes).also {
            for (i in 0 until bitmaskBytes - 1) it[i] = 0x80.toByte()
        }
    val offsetTable = mutableListOf<Int>()
    val slotOffsets = mutableListOf<Pair<Int, Int>>() // (slot, bufPos)
    var lastSlot: Int = -1
    var needsSort: Boolean = false
}

private const val sigilStr = 0x22 // '"' — string / var-length
private const val sigilI64 = 0x69 // 'i'
private const val sigilF64 = 0x64 // 'd'
private const val sigilBool = 0x62 // 'b'
private const val sigilNull = 0x6E // 'n'
private const val sigilBinary = 0x42 // 'B'

class NxsWriter(private val schema: NxsSchema) {
    private val buf = ByteArrayOutputStream(4096)
    private val frames = ArrayDeque<Frame>()
    private val recordOffsets = mutableListOf<Int>()
    private val slotSigils = IntArray(schema.count) { sigilStr }

    fun beginObject() {
        if (frames.isEmpty()) recordOffsets += buf.size()
        val start = buf.size()
        val frame = Frame(start, schema.bitmaskBytes)
        frames.addLast(frame)

        writeU32(0x4E59584F) // NYXO
        writeU32(0) // length placeholder
        buf.write(frame.bitmask)
        repeat(schema.count * 2) { buf.write(0) }
        while ((buf.size() - start) % 8 != 0) buf.write(0)
    }

    fun endObject() {
        check(frames.isNotEmpty()) { "endObject without beginObject" }
        val frame = frames.removeLast()
        val bytes = buf.toByteArray() // needed for back-patching
        // Note: we operate on buf's internal array via a trick —
        // ByteArrayOutputStream doesn't expose the array, so we rebuild.
        // This is acceptable since endObject is called per-record, not per-field.
        val arr = bytes
        val totalLen = arr.size - frame.start

        // Back-patch Length at start + 4
        putU32(arr, frame.start + 4, totalLen)

        // Back-patch bitmask at start + 8
        val bmOff = frame.start + 8
        frame.bitmask.forEachIndexed { i, b -> arr[bmOff + i] = b }

        // Back-patch offset table
        val otStart = bmOff + schema.bitmaskBytes
        val present = frame.offsetTable.size

        if (!frame.needsSort) {
            frame.offsetTable.forEachIndexed { i, rel ->
                putU16(arr, otStart + i * 2, rel)
            }
        } else {
            val sorted = frame.slotOffsets.sortedBy { it.first }
            sorted.forEachIndexed { i, (_, bufPos) ->
                putU16(arr, otStart + i * 2, bufPos - frame.start)
            }
        }
        for (i in present * 2 until schema.count * 2) arr[otStart + i] = 0

        buf.reset()
        buf.write(arr)
    }

    fun finish(): ByteArray {
        check(frames.isEmpty()) { "unclosed objects" }

        val schemaBytes = buildSchemaBytes()
        val dictHash = murmurHash64(schemaBytes)
        val dataStart = 32 + schemaBytes.size
        val dataSector = buf.toByteArray()
        val tailPtr = (dataStart + dataSector.size).toLong()
        val tail = buildTailIndex(dataStart, tailPtr)

        val out = ByteArrayOutputStream(32 + schemaBytes.size + dataSector.size + tail.size)

        fun u32(v: Int) {
            out.write(v and 0xFF)
            out.write((v shr 8) and 0xFF)
            out.write((v shr 16) and 0xFF)
            out.write((v shr 24) and 0xFF)
        }

        fun u16(v: Int) {
            out.write(v and 0xFF)
            out.write((v shr 8) and 0xFF)
        }

        fun u64(v: Long) {
            for (i in 0..7) out.write(((v shr (i * 8)) and 0xFF).toInt())
        }

        u32(0x4E595842) // NYXB
        u16(0x0101) // VERSION
        u16(0x0002) // FLAG_SCHEMA_EMBEDDED
        u64(dictHash)
        u64(0)
        repeat(8) { out.write(0) } // reserved

        out.write(schemaBytes)
        out.write(dataSector)
        out.write(tail)
        return out.toByteArray()
    }

    fun writeI64(
        slot: Int,
        value: Long,
    ) {
        slotSigils[slot] = sigilI64
        markSlot(slot)
        for (i in 0..7) buf.write(((value ushr (i * 8)) and 0xFF).toInt())
    }

    fun writeF64(
        slot: Int,
        value: Double,
    ) {
        slotSigils[slot] = sigilF64
        writeI64(slot, java.lang.Double.doubleToRawLongBits(value))
    }

    fun writeBool(
        slot: Int,
        value: Boolean,
    ) {
        slotSigils[slot] = sigilBool
        markSlot(slot)
        buf.write(if (value) 1 else 0)
        repeat(7) { buf.write(0) }
    }

    fun writeTime(
        slot: Int,
        unixNs: Long,
    ) {
        slotSigils[slot] = sigilI64
        writeI64(slot, unixNs)
    }

    fun writeNull(slot: Int) {
        slotSigils[slot] = sigilNull
        markSlot(slot)
        repeat(8) { buf.write(0) }
    }

    fun writeStr(
        slot: Int,
        value: String,
    ) {
        slotSigils[slot] = sigilStr
        markSlot(slot)
        val b = value.toByteArray(Charsets.UTF_8)
        writeU32(b.size)
        buf.write(b)
        val used = (4 + b.size) % 8
        if (used != 0) repeat(8 - used) { buf.write(0) }
    }

    fun writeBytes(
        slot: Int,
        value: ByteArray,
    ) {
        slotSigils[slot] = sigilBinary
        markSlot(slot)
        writeU32(value.size)
        buf.write(value)
        val used = (4 + value.size) % 8
        if (used != 0) repeat(8 - used) { buf.write(0) }
    }

    fun writeListI64(
        slot: Int,
        values: LongArray,
    ) {
        markSlot(slot)
        val total = 16 + values.size * 8
        writeU32(0x4E59584C)
        writeU32(total)
        buf.write(0x3D) // '=' sigil
        writeU32(values.size)
        repeat(3) { buf.write(0) }
        for (v in values) writeRawI64(v)
    }

    fun writeListF64(
        slot: Int,
        values: DoubleArray,
    ) {
        markSlot(slot)
        val total = 16 + values.size * 8
        writeU32(0x4E59584C)
        writeU32(total)
        buf.write(0x7E) // '~' sigil
        writeU32(values.size)
        repeat(3) { buf.write(0) }
        for (v in values) writeRawI64(java.lang.Double.doubleToRawLongBits(v))
    }

    // Convenience: write records from a list of maps.
    companion object {
        fun fromRecords(
            keys: List<String>,
            records: List<Map<String, Any?>>,
        ): ByteArray {
            val schema = NxsSchema(keys)
            val w = NxsWriter(schema)
            for (rec in records) {
                w.beginObject()
                for ((i, key) in keys.withIndex()) {
                    if (!rec.containsKey(key)) continue
                    when (val v = rec[key]) {
                        null -> w.writeNull(i)
                        is Boolean -> w.writeBool(i, v)
                        is Int -> w.writeI64(i, v.toLong())
                        is Long -> w.writeI64(i, v)
                        is Float -> w.writeF64(i, v.toDouble())
                        is Double -> w.writeF64(i, v)
                        is String -> w.writeStr(i, v)
                        is ByteArray -> w.writeBytes(i, v)
                    }
                }
                w.endObject()
            }
            return w.finish()
        }
    }

    private fun markSlot(slot: Int) {
        check(frames.isNotEmpty()) { "write outside beginObject/endObject" }
        val frame = frames.last()
        frame.bitmask[slot / 7] = (frame.bitmask[slot / 7].toInt() or (1 shl (slot % 7))).toByte()
        val rel = buf.size() - frame.start
        if (slot < frame.lastSlot) frame.needsSort = true
        frame.lastSlot = slot
        frame.offsetTable += rel
        frame.slotOffsets += Pair(slot, buf.size())
    }

    private fun writeU32(v: Int) {
        buf.write(v and 0xFF)
        buf.write((v shr 8) and 0xFF)
        buf.write((v shr 16) and 0xFF)
        buf.write((v shr 24) and 0xFF)
    }

    private fun writeRawI64(v: Long) {
        for (i in 0..7) buf.write(((v ushr (i * 8)) and 0xFF).toInt())
    }

    private fun putU32(
        arr: ByteArray,
        off: Int,
        v: Int,
    ) {
        arr[off] = (v and 0xFF).toByte()
        arr[off + 1] = ((v shr 8) and 0xFF).toByte()
        arr[off + 2] = ((v shr 16) and 0xFF).toByte()
        arr[off + 3] = ((v shr 24) and 0xFF).toByte()
    }

    private fun putU16(
        arr: ByteArray,
        off: Int,
        v: Int,
    ) {
        arr[off] = (v and 0xFF).toByte()
        arr[off + 1] = ((v shr 8) and 0xFF).toByte()
    }

    private fun buildSchemaBytes(): ByteArray {
        val n = schema.count
        val utf8 = schema.keys.map { it.toByteArray(Charsets.UTF_8) }
        var size = 2 + n + utf8.sumOf { it.size + 1 }
        val pad = (8 - size % 8) % 8
        size += pad

        val b = ByteArray(size)
        var p = 0
        b[p++] = (n and 0xFF).toByte()
        b[p++] = ((n shr 8) and 0xFF).toByte()
        repeat(n) { i -> b[p++] = slotSigils[i].toByte() }
        for (e in utf8) {
            e.copyInto(b, p)
            p += e.size
            b[p++] = 0
        }
        return b
    }

    private fun buildTailIndex(
        dataStart: Int,
        tailPtr: Long,
    ): ByteArray {
        val nr = recordOffsets.size
        val out = ByteArrayOutputStream(4 + nr * 10 + 12)

        fun u32(v: Int) {
            out.write(v and 0xFF)
            out.write((v shr 8) and 0xFF)
            out.write((v shr 16) and 0xFF)
            out.write((v shr 24) and 0xFF)
        }

        fun u64(v: Long) {
            for (i in 0..7) out.write(((v ushr (i * 8)) and 0xFF).toInt())
        }

        fun u16(v: Int) {
            out.write(v and 0xFF)
            out.write((v shr 8) and 0xFF)
        }

        u32(nr)
        for ((i, rel) in recordOffsets.withIndex()) {
            u16(i)
            u64((dataStart + rel).toLong())
        }
        u64(tailPtr)
        u32(0x2153584E) // NXS!
        return out.toByteArray()
    }
}
